client/client.go

```diff
@@ -1512,28 +1523,44 @@ OUTER:
 	}
 }
 
-// watchNodeUpdates periodically checks for changes to the node attributes or
-// meta map. The passed hashes are the initial hash values for the attribute and
-// metadata of the node respectively.
-func (c *Client) watchNodeUpdates(attrHash, metaHash uint64) {
-	c.logger.Printf("[DEBUG] client: periodically checking for node changes at duration %v", nodeUpdateRetryIntv)
+// updateNode triggers a client to update its node copy if it isn't doing
+// so already
+func (c *Client) updateNode() {
+	select {
+	case c.triggerNodeUpdate <- struct{}{}:
+		// Node update goroutine was released to execute
+	default:
+		// Node update goroutine was already running
+	}
+}
+
+// watchNodeUpdates blocks until it is edge triggered. Once triggered,
+// it will update the client node copy and re-register the node.
+func (c *Client) watchNodeUpdates() {
+	var hasChanged bool
+	timer := time.NewTimer(c.retryIntv(nodeUpdateRetryIntv))
+	defer timer.Stop()
 
-	var changed bool
 	for {
 		select {
-		case <-time.After(c.retryIntv(nodeUpdateRetryIntv)):
-			changed, attrHash, metaHash = c.hasNodeChanged(attrHash, metaHash)
-			if changed {
-				c.logger.Printf("[DEBUG] client: state changed, updating node.")
+		case <-timer.C:
+			if !hasChanged {
+				continue
+			}
+			c.logger.Printf("[DEBUG] client: state changed, updating node and re-registering.")
 
-				// Update the config copy.
-				c.configLock.Lock()
-				node := c.config.Node.Copy()
-				c.configCopy.Node = node
-				c.configLock.Unlock()
+			// Update the config copy.
+			c.configLock.Lock()
+			node := c.config.Node.Copy()
+			c.configCopy.Node = node
+			c.configLock.Unlock()
 
-				c.retryRegisterNode()
-			}
+			c.retryRegisterNode()
+
+			hasChanged = false
+			timer.Reset(c.retryIntv(nodeUpdateRetryIntv))
+		case <-c.triggerNodeUpdate:
+			hasChanged = true
```

Comment:
Why not put all timer logic in this case and then remove the if statement at the top of the timer case and the reset at the bottom:

```
case <-c.triggerNodeUpdate:
  if hasChanged {
     continue
  }
  hasChanged = true
  timer.Reset(c.retryIntv(nodeUpdateRetryIntv))
```